### PR TITLE
[theplatform] prefer hls streams

### PR DESCRIPTION
**Pull request by remitamine**
this is for 2 reasons.
- it solves the problem happen in many sites that use the platform like the one heppen in #3811 seamusphelan already talk about downloading m3u manifest.
i didn't test it in SBS but i try it in USA Network and FOX and it works(some sites need some tweaks like in usa network the user agent should be the one from curl) if the change accepted i will try to add extractors for fox.com and usanetwork.
i know that it's possible to download from these sites with f4m with more changes(the f4m url should be changed manually if the change doesn't accepted i'll try to fix the platform f4m downloading) but there is also the second reason.
- the hls stream is better in streaming with video players like mpv and vlc.

### CODE REVIEW

**Comment by dstftw** [comment]
This breaks `http://www.history.com/topics/valentines-day/history-of-valentines-day/videos/bet-you-didnt-know-valentines-day?m=528e394da93ae&s=undefined&f=1&free=false` and probably lots of another extractors that all have to be tested before making such change.

**Comment by remitamine** [comment]
i fix it for direct download links in smil manifest.

**Comment by dstftw** [comment]
It may contain more than a single video.

**Comment by remitamine** [comment]
if i can get an example so i can test it.

**Comment by dstftw** [comment]
Even with this fix the general issue (unassigned `formats`) is not solved when all paths failed.

File: youtube_dl/extractor/theplatform.py [hunk 1]
```diff
@@ -144,11 +144,13 @@ def _real_extract(self, url):
...
-        f4m_node = body.find(_x('smil:seq//smil:video'))
-        if f4m_node is None:
-            f4m_node = body.find(_x('smil:seq/smil:video'))
-        if f4m_node is not None and '.f4m' in f4m_node.attrib['src']:
-            f4m_url = f4m_node.attrib['src']
+        node = body.find(_x('smil:seq//smil:video'))
+        if node is None:
+            node = body.find(_x('smil:seq/smil:video'))
+        if node is not None and '.m3u8' in node.attrib['src']:
+            formats = self._extract_m3u8_formats(node.attrib['src'], video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/theplatform.py]: m3u8 can't be extracted now since `formats` will be overwritten by the code below.

**Comment by remitamine** [comment]
i didn't know if that what you mean.
what i understand is that some videos can contain different extractions methods in the same video like in the snagfilms extractor.

**Comment by dstftw** [comment]
No, that wasn't what I mean.
[These formats](https://github.com/remitamine/youtube-dl/blob/aa55903fbe4669c2006fcc2dc1991a3fdce80a3f/youtube_dl/extractor/theplatform.py#L151) will be overwritten in any case by code below [this](https://github.com/remitamine/youtube-dl/blob/aa55903fbe4669c2006fcc2dc1991a3fdce80a3f/youtube_dl/extractor/theplatform.py#L159) or [this](https://github.com/remitamine/youtube-dl/blob/aa55903fbe4669c2006fcc2dc1991a3fdce80a3f/youtube_dl/extractor/theplatform.py#L161).

**Comment by remitamine** [comment]
but i change it to extand so it should add to formats array and not overwrite it.

**Comment by dstftw** [comment]
There is already a [generic solution](https://github.com/rg3/youtube-dl/blob/c6b68648f45498bcacc71cc8f29696fa93259a7a/youtube_dl/extractor/common.py#L971) to extract all formats from smil.

**Comment by remitamine** [comment]
i try to use _extract_smil_formats but it didn't work:
```
youtube-dl -v 'http://www.history.com/topics/valentines-day/history-of-valentines-day/videos/bet-you-didnt-know-valentines-day?m=528e394da93ae&s=undefined&f=1&free=false'
[debug] System config: []
[debug] User config: ['--external-downloader', 'aria2c', '--external-downloader-args', '-x 16 --check-certificate=false', '-f', 'best[height<=?720]', '--sub-lang', 'ar,en', '--write-sub']
[debug] Command-line args: ['-v', 'http://www.history.com/topics/valentines-day/history-of-valentines-day/videos/bet-you-didnt-know-valentines-day?m=528e394da93ae&s=undefined&f=1&free=false']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2015.07.07
[debug] Python version 3.4.3 - Linux-4.1.2-2-ARCH-x86_64-with-arch-Arch-Linux
[debug] exe versions: ffmpeg 2.7.1, ffprobe 2.7.1, rtmpdump 2.4
[debug] Proxy map: {}
[History] bet-you-didnt-know-valentines-day: Downloading webpage
[ThePlatform] bLx5Dv5Aka1G: Downloading SMIL file
Traceback (most recent call last):
  File "/usr/lib/python3.4/runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.4/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "./youtube-dl/__main__.py", line 19, in <module>
  File "./youtube-dl/youtube_dl/__init__.py", line 410, in main
  File "./youtube-dl/youtube_dl/__init__.py", line 400, in _real_main
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 1504, in download
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 667, in extract_info
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 720, in process_ie_result
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 656, in extract_info
  File "./youtube-dl/youtube_dl/extractor/common.py", line 275, in extract
  File "./youtube-dl/youtube_dl/extractor/theplatform.py", line 119, in _real_extract
  File "./youtube-dl/youtube_dl/extractor/common.py", line 979, in _extract_smil_formats
AttributeError: 'NoneType' object has no attribute 'get'
```
it happens in this line:
base = smil.find('./head/meta').get('base')
the header of the smil file contains this:
```
<smil xmlns="http://www.w3.org/2005/SMIL21/Language" xmlns:tp="http://xml.theplatform.com/mps/metadata/content/custom">
<head>
	<meta name="refreshToken" content="014ea11c7aa040ea40c972340bf12bbd77df965c24189132148a2dfe6afcfbe40c9c06370a31"/>
</head>

**Comment by dstftw** [comment]
These changes break at least `https://player.theplatform.com/p/D6x-PC/pulse_preview/embed/select/media/yMBg9E8KFxZD`. If you want this to be merged you have to check all youtube-dl tests and possibly all closed theplatform issue related to theplatform extraction and make sure there are no regressions.

**Comment by remitamine** [comment]
in the video you send the video_id extracted from url is (media) in the branch that i've created and in the upstream/master it extract the real id (yMBg9E8KFxZD).
i don't know why this heppen because i didn't change the _VALID_URL
i run the tests for all the extractors that use theplatform
there are errors because:
- the tests assume that the output is flv but with hls it is m3u8.
- errors in the extractor before reach theplatform extractor
- there are others that don't work because the changes i made i will try to fix them.

**Comment by dstftw** [comment]
You did not get it. You should check that all these videos are extracted properly, downloadable and the resulting quality is comparable to one obtained with current code. Obviously, they will fail now due to diffent `ext` and so.
With this patch https://player.theplatform.com/p/D6x-PC/pulse_preview/embed/select/media/yMBg9E8KFxZD can't be downloaded at all:
```
C:\Dev\git\youtube-dl\master>python -m youtube_dl "https://player.theplatform.co
m/p/D6x-PC/pulse_preview/embed/select/media/yMBg9E8KFxZD" -v
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'https://player.theplatform.com/p/D6x-PC/pulse_prev
iew/embed/select/media/yMBg9E8KFxZD', u'-v']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2015.07.18
[debug] Git HEAD: db02344
[debug] Python version 2.7.10 - Windows-2003Server-5.2.3790-SP2
[debug] exe versions: ffmpeg N-71608-ga924b83, ffprobe N-71608-ga924b83, rtmpdum
p 2.4
[debug] Proxy map: {}
[ThePlatform] yMBg9E8KFxZD: Downloading XML
[ThePlatform] yMBg9E8KFxZD: Downloading webpage
[ThePlatform] yMBg9E8KFxZD: Downloading m3u8 information
[debug] Invoking downloader on u'http://tpr205492def-vh.akamaihd.net/i/,ESM_Rugb
y/122/47/londondaytwohighlights__778794.mp4,.csmil/index_0_av.m3u8?null='
[download] Destination: HIGHLIGHTS - USA bag first ever series Cup win-yMBg9E8KF
xZD.m3u8
ffmpeg version N-71608-ga924b83 Copyright (c) 2000-2015 the FFmpeg developers
  built with gcc 4.9.2 (GCC)
  configuration: --enable-gpl --enable-version3 --disable-w32threads --enable-av
isynth --enable-bzlib --enable-fontconfig --enable-frei0r --enable-gnutls --enab
le-iconv --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --
enable-libdcadec --enable-libfreetype --enable-libgme --enable-libgsm --enable-l
ibilbc --enable-libmodplug --enable-libmp3lame --enable-libopencore-amrnb --enab
le-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-librtmp --en
able-libschroedinger --enable-libsoxr --enable-libspeex --enable-libtheora --ena
ble-libtwolame --enable-libvidstab --enable-libvo-aacenc --enable-libvo-amrwbenc
 --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enabl
e-libx264 --enable-libx265 --enable-libxavs --enable-libxvid --enable-lzma --ena
ble-decklink --enable-zlib
  libavutil      54. 23.101 / 54. 23.101
  libavcodec     56. 35.101 / 56. 35.101
  libavformat    56. 30.100 / 56. 30.100
  libavdevice    56.  4.100 / 56.  4.100
  libavfilter     5. 14.100 /  5. 14.100
  libswscale      3.  1.101 /  3.  1.101
  libswresample   1.  1.100 /  1.  1.100
  libpostproc    53.  3.100 / 53.  3.100
[http @ 0000000002e11300] HTTP error 403 Forbidden
http://tpr205492def-vh.akamaihd.net/i/,ESM_Rugby/122/47/londondaytwohighlights__
778794.mp4,.csmil/index_0_av.m3u8?null=: Server returned 403 Forbidden (access d
enied)


ERROR: ffmpeg exited with code 1
  File "C:\Python\Python2710\lib\runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "C:\Python\Python2710\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Dev\git\youtube-dl\master\youtube_dl\__main__.py", line 19, in <modul
e>
    youtube_dl.main()
  File "youtube_dl\__init__.py", line 410, in main
    _real_main(argv)
  File "youtube_dl\__init__.py", line 400, in _real_main
    retcode = ydl.download(all_urls)
  File "youtube_dl\YoutubeDL.py", line 1503, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', Fals
e))
  File "youtube_dl\YoutubeDL.py", line 667, in extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "youtube_dl\YoutubeDL.py", line 713, in process_ie_result
    return self.process_video_result(ie_result, download=download)
  File "youtube_dl\YoutubeDL.py", line 1173, in process_video_result
    self.process_info(new_info)
  File "youtube_dl\YoutubeDL.py", line 1435, in process_info
    success = dl(filename, info_dict)
  File "youtube_dl\YoutubeDL.py", line 1377, in dl
    return fd.download(name, info)
  File "youtube_dl\downloader\common.py", line 342, in download
    return self.real_download(filename, info_dict)
  File "youtube_dl\downloader\hls.py", line 50, in real_download
    self.report_error('%s exited with code %d' % (ffpp.basename, retval))
  File "youtube_dl\downloader\common.py", line 155, in report_error
    self.ydl.report_error(*args, **kargs)
  File "youtube_dl\YoutubeDL.py", line 531, in report_error
    self.trouble(error_message, tb)
  File "youtube_dl\YoutubeDL.py", line 493, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())
```

**Comment by remitamine** [comment]
some of the videos need a different header for example in the videos from USA Network when i try to download from it manualy using the m3u8 i should change the user agent to 'curl/7.34.0' to if i use the default user agent it wont work i will test and see the different configuration

**Comment by remitamine** [comment]
i fix the test for:
- NationalGeographic
- ThePplatform
- FoxSports
- CNET
- SBS
- NBC
test still didn't pass because of the changes in this branch:
- CBS (i will fix it)
test didn't pass because they are broken(not related to the changes this branch):
- CBSSports
```
[CBSSports] 318462531970: Downloading JSON metadata
ERROR: An extractor error has occured. (caused by StopIteration());
```
- Syfy
```
[Syfy] 2631458: Downloading webpage
ERROR: Unable to extract video url;
```

**Comment by dstftw** [comment]
Can you name an example URL that depends on headers/cookies for final m3u8?

File: youtube_dl/downloader/hls.py [hunk 2]
```diff
@@ -28,9 +28,12 @@ def real_download(self, filename, info_dict):
             return False
         ffpp.check_version()
 
+        cookies = ''
+        for cookie in self.ydl.cookiejar:
+            cookies += '%s=%s; path=%s; domain=%s;\r\n' % (cookie.name, cookie.value, cookie.path, cookie.domain)
         args = [
             encodeArgument(opt)
-            for opt in (ffpp.executable, '-y', '-i', url, '-f', 'mp4', '-c', 'copy', '-bsf:a', 'aac_adtstoasc')]
+            for opt in (ffpp.executable, '-y', '-user-agent', info_dict['http_headers']['User-Agent'], '-cookies', cookies,'-i', url, '-f', 'mp4', '-c', 'copy', '-bsf:a', 'aac_adtstoasc')]
```
Comment (dstftw) [inline youtube_dl/downloader/hls.py]: avconv does not support `-cookies`, use `-headers` instead. You should also pass all headers.

Comment (remitamine) [inline youtube_dl/downloader/hls.py]: for example the url you send before:
before the commit:
```
youtube-dl https://player.theplatform.com/p/D6x-PC/pulse_preview/embed/select/media/yMBg9E8KFxZD
[ThePlatform] yMBg9E8KFxZD: Downloading XML
[ThePlatform] yMBg9E8KFxZD: Downloading webpage
[ThePlatform] yMBg9E8KFxZD: Downloading m3u8 information
[download] Destination: HIGHLIGHTS - USA bag first ever series Cup win-yMBg9E8KFxZD.m3u8
ffmpeg version 2.7.2 Copyright (c) 2000-2015 the FFmpeg developers
  built with gcc 5.1.0 (GCC)
  configuration: --prefix=/usr --disable-debug --disable-static --disable-stripping --enable-avisynth --enable-avresample --enable-fontconfig --enable-gnutls --enable-gpl --enable-libass --enable-libbluray --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-libschroedinger --enable-libspeex --enable-libssh --enable-libtheora --enable-libv4l2 --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxvid --enable-shared --enable-version3 --enable-x11grab
  libavutil      54. 27.100 / 54. 27.100
  libavcodec     56. 41.100 / 56. 41.100
  libavformat    56. 36.100 / 56. 36.100
  libavdevice    56.  4.100 / 56.  4.100
  libavfilter     5. 16.101 /  5. 16.101
  libavresample   2.  1.  0 /  2.  1.  0
  libswscale      3.  1.101 /  3.  1.101
  libswresample   1.  2.100 /  1.  2.100
  libpostproc    53.  3.100 / 53.  3.100
[http @ 0x5622e2c72b80] HTTP error 403 Forbidden
http://tpr205492def-vh.akamaihd.net/i/,ESM_Rugby/122/47/londondaytwohighlights__778794.mp4,.csmil/index_0_av.m3u8?null=: Server returned 403 Forbidden (access denied)


ERROR: ffmpeg exited with code 1
```
after the commit:
```
https://player.theplatform.com/p/D6x-PC/pulse_preview/embed/select/media/yMBg9E8KFxZD
[ThePlatform] yMBg9E8KFxZD: Downloading XML
[ThePlatform] yMBg9E8KFxZD: Downloading webpage
[ThePlatform] yMBg9E8KFxZD: Downloading m3u8 information
[download] Destination: HIGHLIGHTS - USA bag first ever series Cup win-yMBg9E8KFxZD.m3u8
ffmpeg version 2.7.2 Copyright (c) 2000-2015 the FFmpeg developers
  built with gcc 5.1.0 (GCC)
  configuration: --prefix=/usr --disable-debug --disable-static --disable-stripping --enable-avisynth --enable-avresample --enable-fontconfig --enable-gnutls --enable-gpl --enable-libass --enable-libbluray --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-libschroedinger --enable-libspeex --enable-libssh --enable-libtheora --enable-libv4l2 --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxvid --enable-shared --enable-version3 --enable-x11grab
  libavutil      54. 27.100 / 54. 27.100
  libavcodec     56. 41.100 / 56. 41.100
  libavformat    56. 36.100 / 56. 36.100
  libavdevice    56.  4.100 / 56.  4.100
  libavfilter     5. 16.101 /  5. 16.101
  libavresample   2.  1.  0 /  2.  1.  0
  libswscale      3.  1.101 /  3.  1.101
  libswresample   1.  2.100 /  1.  2.100
  libpostproc    53.  3.100 / 53.  3.100
[h264 @ 0x562900f6a240] non-existing SPS 0 referenced in buffering period
    Last message repeated 1 times
[h264 @ 0x562901025300] non-existing SPS 0 referenced in buffering period
Input #0, hls,applehttp, from 'http://tpr205492def-vh.akamaihd.net/i/,ESM_Rugby/122/47/londondaytwohighlights__778794.mp4,.csmil/index_0_av.m3u8?null=':
  Duration: 00:07:36.00, start: 0.100511, bitrate: 0 kb/s
  Program 0 
    Metadata:
      variant_bitrate : 0
    Stream #0:0: Video: h264 (High) ([27][0][0][0] / 0x001B), yuv420p(tv, bt709), 1280x720 [SAR 1:1 DAR 16:9], 25 fps, 25 tbr, 90k tbn, 50 tbc
    Stream #0:1: Audio: aac (LC) ([15][0][0][0] / 0x000F), 44100 Hz, stereo, fltp, 129 kb/s
    Stream #0:2: Data: timed_id3 (ID3  / 0x20334449)
[mp4 @ 0x562901005c00] Codec for stream 0 does not use global headers but container format requires global headers
[mp4 @ 0x562901005c00] Codec for stream 1 does not use global headers but container format requires global headers
Output #0, mp4, to 'HIGHLIGHTS - USA bag first ever series Cup win-yMBg9E8KFxZD.m3u8.part':
  Metadata:
    encoder         : Lavf56.36.100
    Stream #0:0: Video: h264 ([33][0][0][0] / 0x0021), yuv420p, 1280x720 [SAR 1:1 DAR 16:9], q=2-31, 25 fps, 25 tbr, 90k tbn, 90k tbc
    Stream #0:1: Audio: aac ([64][0][0][0] / 0x0040), 44100 Hz, stereo, 129 kb/s
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
  Stream #0:1 -> #0:1 (copy)
Press [q] to stop, [?] for help
frame=   23 fps=0.0 q=-1.0 size=     148kB time=00:00:00.90 bitrate=1336.1kbits/frame=   25 fps= 19 q=-1.0 size=     169kB time=00:00:00.99 bitrate=1383.6kbits/
```

Comment (remitamine) [inline youtube_dl/downloader/hls.py]: i send only the cookies and user agent because they are the only information that i find some sites test for

**Comment by dstftw** [comment]
Why did you close this?

**Comment by remitamine** [comment]
i think may be i'm wrong.
it better to extract both f4m and m3u8 and give the user the chance to choose between them and not force him to use one of them if i have time i will try to make it or may another one.

**Comment by dstftw** [comment]
I've incorporated passing http headers to hls downloader code from c8c847d4478484e51a2f5294682711741b854c2f in f1028194636e7acafb29fb38244cc7b1347d9313, thanks.

**Pull request closed without merging**